Chernoff distribution

NittanyLion opened the pull request:
I've added code to compute the Chernoff distribution; see chernoff.jl header for references.

Since this is my first submission, I'm sure that there are a number of infelicities in terms of integration and such.  Please advise.

dmbates commented:
- It looks as if you are using an airyprime function in `chernoff_p`.  If so it would be better to see if you can get the value from a function in the [`SpecialFunctions` package](https://github.com/JuliaMath/SpecialFunctions.jl), which is already being used by this package.

- the standard indentation in Julia packages is four spaces

- generally we would add unit tests (in the `test` subdirectory) of the various functions when adding a distribution.

- note the style of documentation in other source files.  The quoted blocks (blocks beginning and ending with lines consisting only of triple double-quote characters) are incorporated into the package documentation.

- Julia has a cube-root function, `cbrt`

- checks for `x == 0` can be written `iszero(x)` 

- I am happy to respond to questions, via email if you wish, on how to change this PR.

NittanyLion commented:
Thanks Douglas.

Ok, I'll make those changes when I get a chance.  Will take a little
while.  Meanwhile, should I cancel the pull request?

Joris


On Wed, Dec 13, 2017 at 4:35 PM, Douglas Bates <notifications@github.com>
wrote:

>
>    -
>
>    It looks as if you are using an airyprime function in chernoff_p. If
>    so it would be better to see if you can get the value from a function in
>    the SpecialFunctions package
>    <https://github.com/JuliaMath/SpecialFunctions.jl>, which is already
>    being used by this package.
>    -
>
>    the standard indentation in Julia packages is four spaces
>    -
>
>    generally we would add unit tests (in the test subdirectory) of the
>    various functions when adding a distribution.
>    -
>
>    note the style of documentation in other source files. The quoted
>    blocks (blocks beginning and ending with lines consisting only of triple
>    double-quote characters) are incorporated into the package documentation.
>    -
>
>    Julia has a cube-root function, cbrt
>    -
>
>    checks for x == 0 can be written iszero(x)
>    -
>
>    I am happy to respond to questions, via email if you wish, on how to
>    change this PR.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaStats/Distributions.jl/pull/695#issuecomment-351532482>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APtv5qnIaWEWlRsl9VCY5XtTailaDVTiks5tAEM2gaJpZM4RBKiW>
> .
>

ararslan commented:
No need to cancel the pull request. You can just push your changes to this branch as you make them and the pull request will be updated.

NittanyLion commented:
As an aside, the airyprime references are simply constants that I
precomputed using the SpecialFunctions package to improve computation
speed.  There is no computation of airy functions or their derivatives.


On Wed, Dec 13, 2017 at 4:35 PM, Douglas Bates <notifications@github.com>
wrote:

>
>    -
>
>    It looks as if you are using an airyprime function in chernoff_p. If
>    so it would be better to see if you can get the value from a function in
>    the SpecialFunctions package
>    <https://github.com/JuliaMath/SpecialFunctions.jl>, which is already
>    being used by this package.
>    -
>
>    the standard indentation in Julia packages is four spaces
>    -
>
>    generally we would add unit tests (in the test subdirectory) of the
>    various functions when adding a distribution.
>    -
>
>    note the style of documentation in other source files. The quoted
>    blocks (blocks beginning and ending with lines consisting only of triple
>    double-quote characters) are incorporated into the package documentation.
>    -
>
>    Julia has a cube-root function, cbrt
>    -
>
>    checks for x == 0 can be written iszero(x)
>    -
>
>    I am happy to respond to questions, via email if you wish, on how to
>    change this PR.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaStats/Distributions.jl/pull/695#issuecomment-351532482>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APtv5qnIaWEWlRsl9VCY5XtTailaDVTiks5tAEM2gaJpZM4RBKiW>
> .
>

NittanyLion commented:
Ok, I have uploaded my changes.  There are significant changes to the chernoff.jl source file, I've added a chernoff.jl test file, and made changes to src/univariates.jl, test/runtests.jl, and src/Distributions.jl .  Pkg.test("Distributions") runs fine.  I've incorporated all suggested changes.

The only trouble I had was with uploading / the github system.  I ended up using the web interface to make changes, which is suboptimal.  I'm not sure what went wrong.

NittanyLion commented:
Ok, I have uploaded my changes. There are significant changes to the
chernoff.jl source file, I've added a chernoff.jl test file, and made
changes to src/univariates.jl, test/runtests.jl, and src/Distributions.jl .
Pkg.test("Distributions") runs fine. I've incorporated all suggested
changes.

The only trouble I had was with uploading / the github system. I ended up
using the web interface to make changes, which is suboptimal since it is
more likely that I'll make a mistake. I'm not sure what went wrong

On Wed, Dec 13, 2017 at 4:45 PM, Alex Arslan <notifications@github.com>
wrote:

> No need to cancel the pull request. You can just push your changes to this
> branch as you make them and the pull request will be updated.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaStats/Distributions.jl/pull/695#issuecomment-351535024>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APtv5neOTAWe1zOmNJrORgcrijByVKcuks5tAEWHgaJpZM4RBKiW>
> .
>

ararslan commented on src/samplers/categorical.jl:
```diff
@@ -25,10 +25,16 @@ end
+    struct AliasTable <: Sampleable{Univariate,Discrete}
+        accept::Vector{Float64}
+        alias::Vector{Int}
+        isampler::SamplerRangeInt{Int,UInt}
```
Both of these blocks will need to be terminated with `end`

NittanyLion replied:
Thanks.  Still doesn't compile in v0.7 since it still wants to know the definition of RangeGeneratorInt{Int,UInt} in v0.7, which makes sense.  Clearly have to call it a day.

NittanyLion commented:
The code compiles fine on 0.6.  It fails on nightly due to an issue unrelated to the Distributions package.  What, if anything, should I do now?

matbesancon commented:
@NittanyLion it seems this PR has unfortunately been forgotten for a while, would you mine merging master to update and fix conflicts?

matbesancon commented:
bumping this, we need to close or finish it

NittanyLion commented:
Ok, will do when I get a chance.  Hope to have time around Christmas.

On Sun, Nov 4, 2018, 11:14 AM Mathieu Besançon <notifications@github.com
wrote:

> bumping this, we need to close or finish it
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaStats/Distributions.jl/pull/695#issuecomment-435682764>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APtv5tVRACPO3JN5sBVbwTlvQqCvBfggks5urxJUgaJpZM4RBKiW>
> .
>

matbesancon commented:
ping @NittanyLion

eulerkochy commented:
Can I carry forward this PR? Looks like a lot of work has already been done, it would be a waste of human effort to let all of this go in vain!

matbesancon commented:
Yes, please do :)

eulerkochy commented:
On a first glance, the code seems badly formated, and there have been un-necessary `import`s. Will try to go through the code, and start working on a fresh PR. Is that alright?

matbesancon commented:
Both are ok. If it's just about formatting and imports, starting from this
base seems reasonable.

On Wed, Mar 6, 2019, 12:53 Koustav Chowdhury <notifications@github.com>
wrote:

> On a first glance, the code seems badly formated, and there have been
> un-necessary imports. Will try to go through the code, and start working
> on a fresh PR. Is that alright?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaStats/Distributions.jl/pull/695#issuecomment-470079655>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHRRsuG7pe4cJC_s5I5aJV9m5LwRw8sdks5vT6w8gaJpZM4RBKiW>
> .
>

eulerkochy commented:
But I can't commit on this branch, can I? I don't have the write-access

eulerkochy commented:
Well, I am unable to write to this PR, I tried but it returns `403`, which implies `Forbidden error`. Guess, I'll make a new PR.

matbesancon commented:
You can fork the PR itself

matbesancon commented:
Maybe not optimal, but what I do is:
On my dev version, add a remote `fork` with the other person's url
Fetch the branch from this one, and then work from there. Once you're done, you can push the result to another PR

eulerkochy commented:
Well, I was doing by that method only, but it's returning 403. :(

eulerkochy commented:
Ok, I get you. I'll 'push' the result to a different PR, I was missing that!

matbesancon commented:
Closing this in favour of #840
